### PR TITLE
DSL2: Start adding fastp/adapterremoval

### PR DESCRIPTION
**Pull request by jfy133**
Start adding FASTP/ADAPTERREMOVAL subworkflwos

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
# TODO

- [ ] Complete documentation
- [ ] Add support for all eager2 parameters
- [ ] Add CI tests for AR2/FASTP
- [ ] Check output files published correctly

**Comment by jfy133** [comment]
- To test
    - [ ] adapterlist passed to fastp
    - [ ] adapterlist passed to adapterremoval
    - [ ] adapter sequence passed to fastp
    - [ ] adapter sequenced passed to adapterremoval
    - [ ]  merging executes correctly for fastp (note via module)
    - [ ] merging executes correctly for adapterremoval
    - [ ] check preserve5p included correctly for adapterremoval
    - [ ] check polyg trimming executed correctly for fastp
    - [ ] check (skipping) quality trimming works correctly adapterremoval
    - [ ] check 5p/3p trimming works correctly fastp
    - [ ] check 5p/3p trimming works correctly adapterremoval

**Comment by jfy133** [comment]
Note: 

- Missing MultiQC docs (will do that later, still considering how we should integrate them into the new style of nf-core docs
- Includes the bwa/bowtie modules accidently but will come in next PR anyway!

File: docs/output.md [hunk 1]
```diff
@@ -6,37 +6,40 @@ This document describes the output produced by the pipeline. Most of the plots a
...
-<!-- TODO nf-core: Write this documentation describing your workflow's output -->
-
 ## Pipeline overview
 
 The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:
 
 - [FastQC](#fastqc) - Raw read QC
 - [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
 - [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
+- ]
```
Comment (TCLamnidis) [inline docs/output.md:16]: That's a typo, right?
```suggestion
```

File: docs/output.md [hunk 2]
```diff
@@ -53,6 +56,46 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+### AdapterRemoval
+
+<details markdown="1">
+<summary>Output files</summary>
+
+- `adapterremoval/`
+  - `*.settings`: Log file of the AdapterRemoval execution containing trimming, merging, and other read summary statistics.
+  - `*.truncated.fastq.gz`: Final FASTQ file of single-end data that have undergone quality trimming. These are the final reads used downstream for single-end data.
+  - `*.merged.fastq.gz`: Final FASTQ file of paired-end data that have undergone quality trimming, and merging. Typically a concatenated file of the `*collapsed*` files. May, or may not include unmerged reads (singletons), depending on your parameters. These are the final reads used downstream for paired-end data.
+  - `*.pair{1,2}.truncated.fastq.gz`: Final FASTQ file(s) containing paired-end reads that have had undergone additional quality trimming but _not_ merging. These are the final reads used downstream for paired-end data when merging is skipped. Only present for paired-end libraries.
```
Comment (TCLamnidis) [inline docs/output.md:68]:
```suggestion
  - `*.pair{1,2}.truncated.fastq.gz`: Final FASTQ file(s) containing paired-end reads that have undergone additional quality trimming but _not_ merging. These are the final reads used downstream for paired-end data when merging is skipped. Only present for paired-end libraries.
```

File: docs/output.md [hunk 3]
```diff
@@ -53,6 +56,46 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+  - `*_collapsed.fastq.gz`: FASTQ file containing merged reads of paired-end libraries that have not had trimming performed.
+  - `*_collapsed.truncated.fastq.gz`: FASTQ file containing merged reads that have had additional quality trimming performed.
+  - `*_discarded.fastq.gz`: FASTQ file containing merged reads that did not pass other quality thresholds (e.g. minimum length).
+  - `*.truncated.fastq.gz`: FASTQ file(s) containing umerged reads that have had undergone additional quality trimming. Only present for paired-end libraries.
```
Comment (TCLamnidis) [inline docs/output.md:72]:
```suggestion
  - `*.truncated.fastq.gz`: FASTQ file(s) containing umerged reads that have undergone additional quality trimming. Only present for paired-end libraries.
```

File: docs/output.md [hunk 4]
```diff
@@ -53,6 +56,46 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+### fastp
+
+<details markdown="1">
+<summary>Output files</summary>
+
+- `fastp/`
+
+  - `*.{log,html,json}`: Log files of the AdapterRemoval execution containing trimming, merging, and other read summary statistics. The three files contain the same information stored in different formats.
+  - `*.merged.fastq.gz`: Final FASTQ file of paired-end data that have undergone quality trimming, and merging. May, or may not include unmerged reads (singletons), depending on your parameters. These are the final reads used downstream for paired-end data.
+  - `fastp.fastq.gz`: Final FASTQ file of single-end data that have undergone quality trimming. These are the final reads used downstream for single-end data.
```
Comment (TCLamnidis) [inline docs/output.md:90]:
```suggestion
  - `*.fastp.fastq.gz`: Final FASTQ file of single-end data that have undergone quality trimming. These are the final reads used downstream for single-end data.
```

File: docs/output.md [hunk 5]
```diff
@@ -53,6 +56,46 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+  - `{1,2}.fastp.fastq.gz`: Final FASTQ file(s) containing paired-end reads that have had undergone additional quality trimming but _not_ merging. These are the final reads used downstream for paired-end data when merging is skipped. Only present for paired-end libraries.
```
Comment (TCLamnidis) [inline docs/output.md:91]:
```suggestion
  - `*_{1,2}.fastp.fastq.gz`: Final FASTQ file(s) containing paired-end reads that have had undergone additional quality trimming but _not_ merging. These are the final reads used downstream for paired-end data when merging is skipped. Only present for paired-end libraries.
```

File: nextflow_schema.json [hunk 6]
```diff
@@ -296,6 +296,146 @@
...
+            "type": "object",
+            "description": "Removal of adapters, paired-end merging, poly-G removal etc.",
+            "default": "",
+            "properties": {
+                "skip_preprocessing": {
+                    "type": "boolean",
+                    "description": "Specify to skip all preprocessing steps (adapter removal, paired-end merging, poly-g trimming etc).",
+                    "help_text": "Specify to skip all preprocessing steps (adapter removal, paired-end merging, poly-g trimming etc).\n\nThis will also mean you will only get one set of FastQC results (of the input reads).",
+                    "fa_icon": "fas fa-forward"
+                },
```
Comment (TCLamnidis) [inline nextflow_schema.json:311]: Does this not require a default value of `false`? Is the default of the section (`""`)  enough?

File: nextflow.config [hunk 7]
```diff
@@ -62,6 +62,29 @@ params {
...
+    preprocessing_trim3p                             = 0      // WARN: slightly different behaviour between fastp & adapterremoval
+    preprocessing_fastp_complexityfilter             = false
+    preprocessing_fastp_complexityfilter_threshold   = 10
+    preprocessing_adapterremoval_preserve5p          = false
+    preprocessing_adapterremoval_adapteroverlap      = 1
+    preprocessing_adapterremoval_skipqualitytrimming = false
+    preprocessing_adapterremoval_trimbasequalitymin  = 20
+    preprocessing_adapterremoval_skipntrimming       = false
+    preprocessing_adapterremoval_qualitymax          = 41
+
```
Comment (TCLamnidis) [inline nextflow.config:87]: Check and clear TODOs ploz

File: subworkflows/local/preprocessing_adapterremoval.nf [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    ADAPTERREMOVAL_PAIRED ( ch_input_for_adapterremoval.paired, adapterlist )
+    ch_versions = ch_versions.mix( ADAPTERREMOVAL_PAIRED.out.versions.first() )
+    ch_multiqc_files = ch_multiqc_files.mix(ADAPTERREMOVAL_PAIRED.out.settings)
+
+    /*
+     * Due to the ugly output implementation of the current AdapterRemoval2 version, each file
+     * has to be exported in a separate channel and we must manually recombine when necessary.
+     */
+
+     // TODO Compare against eager2, I have a feeling we are missing the `--preserve5p` linking
```
Comment (TCLamnidis) [inline subworkflows/local/preprocessing_adapterremoval.nf:38]: Do OTODO and clear TODO bitte!

File: workflows/eager.nf [hunk 9]
```diff
@@ -64,18 +69,30 @@ def multiqc_report = []
 
 workflow EAGER {
 
+    log.info "Schaffa, Schaffa, Genome Baua!"
```
Comment (TCLamnidis) [inline workflows/eager.nf:72]: ❤️

**Review by TCLamnidis — CHANGES_REQUESTED** [review]

**Comment by jfy133** [comment]
Typos fixed and relevant TODOs removed! Thanks @TCLamnidis! Feel free to merge if you're happy

**Review by apeltzer — APPROVED** [review]
Looks good to me btu will not merge myself ;-) 🤠

File: nextflow_schema.json (hunk 6 above)
Comment (jfy133) [reply nextflow_schema.json:311]: Apparently default in booelan is assumed to be false 🤷‍♀ I've just rebuilt the schema and same applies (and also checked other pipelines)

**Review by TCLamnidis — APPROVED** [review]

**Pull request merged**